server/src/main/java/org/opensearch/cluster/routing/allocation/allocator/RemoteShardsBalancer.java

```diff
@@ -0,0 +1,598 @@
+ * for remote shards within the cluster.
+ *
+ * @opensearch.internal
+ */
+public final class RemoteShardsBalancer extends ShardsBalancer {
+    private final Logger logger;
+    private final RoutingAllocation allocation;
+    private final RoutingNodes routingNodes;
+
+    public RemoteShardsBalancer(Logger logger, RoutingAllocation allocation) {
+        this.logger = logger;
+        this.allocation = allocation;
+        this.routingNodes = allocation.routingNodes();
+    }
+
+    /**
+     * Allocates unassigned remote shards on the routing node which are filtered using
+     * {@link #groupUnassignedShardsByIndex}
+     */
+    @Override
+    void allocateUnassigned() {
+        if (routingNodes.unassigned().isEmpty()) {
+            logger.debug("No unassigned remote shards found.");
+            return;
+        }
+
+        Queue<RoutingNode> nodeQueue = getShuffledRemoteNodes();
+        if (nodeQueue.isEmpty()) {
+            logger.debug("No remote searcher nodes available for unassigned remote shards.");
+            failUnattemptedShards();
+            return;
+        }
+
+        Map<String, UnassignedIndexShards> unassignedShardMap = groupUnassignedShardsByIndex();
+        allocateUnassignedPrimaries(nodeQueue, unassignedShardMap);
+        allocateUnassignedReplicas(nodeQueue, unassignedShardMap);
+        ignoreRemainingShards(unassignedShardMap);
+    }
+
+    /**
+     * Performs shard movement for incompatible remote shards
+     */
+    @Override
+    void moveShards() {
+        Queue<RoutingNode> eligibleNodes = new ArrayDeque<>();
+        Queue<RoutingNode> excludedNodes = new ArrayDeque<>();
+
+        Iterator<RoutingNode> routingNodesIter = routingNodes.iterator();
+        int throttledNodeCount = 0;
+        while (routingNodesIter.hasNext()) {
+            RoutingNode node = routingNodesIter.next();
+            Decision nodeDecision = allocation.deciders().canAllocateAnyShardToNode(node, allocation);
+            /* canAllocateAnyShardToNode decision can be THROTTLE for throttled nodes. To classify
+             * as excluded nodes, we look for Decision.Type.NO
+             */
+            if (nodeDecision.type() == Decision.Type.NO) {
+                excludedNodes.add(node);
+            } else if (nodeDecision.type() == Decision.Type.YES) {
+                eligibleNodes.add(node);
+            } else {
+                throttledNodeCount++;
+            }
+            logger.debug(
+                "Excluded Node Count: [{}], Eligible Node Count: [{}], Throttled Node Count: [{}]",
+                excludedNodes.size(),
+                eligibleNodes.size(),
+                throttledNodeCount
+            );
+        }
+
+        if (excludedNodes.isEmpty()) {
+            logger.debug("No excluded nodes found. Returning...");
+            return;
+        }
+
+        while (!eligibleNodes.isEmpty() && !excludedNodes.isEmpty()) {
```

Comment:
Is there a way we could simplify this code? While under a for under a while looks to be of higher time complexity, although I am not sure if there's another way to simplify it.

Reply:
This was pointed out as one of the misses within the design itself. Currently we are looking to re-use the correctness of the existing approach.
We will work on optimizing this as a part of phase 2. 
For more references: #4759